tool2schema/schema.py

```diff
@@ -75,6 +75,104 @@ def SaveGPTEnabled(module: ModuleType, path: str, schema_type: SchemaType = Sche
+        an exception is raised if any hallucinated arguments are found
+    :return: A tuple consisting of the function and a dictionary of argument values, or None if no
+        function with the expected name and the `GPTEnabled` decorator is found
+    :raises ParseException: If the argument string is not valid, meaning it is not parsable as JSON,
+        it can be parsed but is not a dictionary, it does not include all required arguments, it
+        includes values that are not of the expected type, or it contains hallucinated arguments
+        and `ignore_hallucinations` is false
+    """
+
+    if not (name := function.get("name", None)):
+        raise ParseException("Function name is missing from the dictionary")
+
+    if not (arguments := function.get("arguments", None)):
+        raise ParseException("Function arguments are missing from the dictionary")
+
+    try:
+        arguments = json.loads(arguments)
+
+    except json.decoder.JSONDecodeError:
+        raise ParseException("Arguments are not in valid JSON format")
+
+    if type(arguments) is not dict:
+        raise ParseException("Arguments are not in the form of a dictionary")
+
+    f = FindGPTEnabledByName(module, name)
+
+    if not f:
+        # A function with the given name was not found
+        return None
+
+    if validate:
+        arguments = _validate_arguments(f, arguments, ignore_hallucinations)
+
+    return f, arguments
+
+
+def _validate_arguments(f: Callable, arguments: dict, ignore_hallucinations: bool) -> dict:
+    """
+    Verify that all required arguments are present, and the arguments are of the expected type.
+    Raise an exception if any of these conditions is not met, or if there are any hallucinated
+    arguments and `ignore_hallucinations` is false.
+
+    :param f: A GPTEnabled-decorated function
+    :param arguments: Arguments to validate
+    :param ignore_hallucinations: Whether to ignore hallucinated arguments or throw an exception
+        if any are present
+    :return: A dictionary of validated arguments
+    """
+    validated = {}
+
+    for key, param in f.schema.parameter_schemas.items():
+        value = arguments.pop(key, Parameter.empty)
+
+        if value == Parameter.empty:
+            # The parameter is missing from the arguments
+            if param.parameter.default == Parameter.empty:
+                # The parameter does not have a default value
+                raise ParseException(
+                    f"Parameter {key} is missing from the arguments "
+                    f"and does not have a default value"
+                )
+        else:
+            if not param.type_schema.validate(value):
+                raise ParseException(f"Argument {key} cannot accept value {value}")
+
+            validated[key] = value
+
+    if not ignore_hallucinations and arguments:
+        raise ParseException(f"Hallucinated argument: {arguments}")
```

Comment:
This is probably an easier to parse string when you have multiple keys?
```suggestion
        raise ParseException(f"Hallucinated argument(s): {', '.join(arguments.keys())}")
```